File: go/internal/display/display.go
// Package display provides display settings and color profile management.
package display

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

func getSetting(serial, namespace, key string) string {
	stdout, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", "settings", "get", namespace, key})
	val := strings.TrimSpace(strings.TrimRight(stdout, "\r\n"))
	if val == "null" || val == "null\r" {
		return ""
	}
	return val
}

func putSetting(serial, namespace, key, value string) error {
	_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell", "settings", "put", namespace, key, value})
	if code != 0 {
		return nterrors.AdbError(fmt.Sprintf("settings put %s %s failed: %s", namespace, key, strings.TrimSpace(stderr)))
	}
	return nil
}

func fmtTimeout(msStr string) string {
	var ms int
	if _, err := fmt.Sscanf(msStr, "%d", &ms); err != nil {
		if msStr != "" {
			return msStr
		}
		return "n/a"
	}
	seconds := ms / 1000
	if seconds < 60 {
		return fmt.Sprintf("%ds", seconds)
	}
	minutes := seconds / 60
	if minutes == 1 {
		return "1 min"
	}
	return fmt.Sprintf("%d min", minutes)
}

func fmtRotation(val string) string {
	m := map[string]string{"0": "portrait", "1": "landscape", "2": "reverse-portrait", "3": "reverse-landscape"}
	if l, ok := m[val]; ok {
		return l
	}
	if val == "" {
		return "n/a"
	}
	return val
}

func fmtOnOff(val string) string {
	if val == "1" {
		return "on"
	}
	if val == "0" {
		return "off"
	}
	if val == "" {
		return "n/a"
	}
	return val
}

var wmSizePhysRe = regexp.MustCompile(`Physical size:\s*(\d+x\d+)`)
var wmSizeRe = regexp.MustCompile(`(\d+x\d+)`)
var wmDensityRe = regexp.MustCompile(`(?:Override|Physical) density:\s*(\d+)`)
var wmDensitySimpleRe = regexp.MustCompile(`(\d+)`)
var refreshRateRe = regexp.MustCompile(`(?:mRefreshRate|refreshRate)[=:\s"]+([0-9]+(?:\.[0-9]+)?)`)

func parseWMSize(raw string) string {
	if m := wmSizePhysRe.FindStringSubmatch(raw); m != nil {
		return m[1] + " (physical)"
	}
	if m := wmSizeRe.FindStringSubmatch(raw); m != nil {
		return m[1]
	}
	if raw == "" {
		return "n/a"
	}
	return raw
}

func parseWMDensity(raw string) string {
	if m := wmDensityRe.FindStringSubmatch(raw); m != nil {
		return m[1]
	}
	if m := wmDensitySimpleRe.FindStringSubmatch(raw); m != nil {
		return m[1]
	}
	if raw == "" {
		return "n/a"
	}
	return raw
}

func parseRefreshRate(raw string) string {
	for _, line := range strings.Split(raw, "\n") {
		if m := refreshRateRe.FindStringSubmatch(line); m != nil {
			return m[1] + " Hz"
		}
	}
	return "n/a"
}

func colorProfileLabel(val string) string {
	m := map[string]string{"0": "Natural (sRGB)", "1": "Vivid (P3)", "256": "Custom"}
	if l, ok := m[val]; ok {
		return l
	}
	if val == "" {
		return "n/a"
	}
	return fmt.Sprintf("Unknown (%s)", val)
}

var displaySettings = map[string][2]string{
	"brightness":      {"system", "screen_brightness"},
	"brightness_auto": {"system", "screen_brightness_mode"},
	"timeout":         {"system", "screen_off_timeout"},
	"rotation":        {"system", "user_rotation"},
	"rotation_auto":   {"system", "accelerometer_rotation"},
	"font_scale":      {"system", "font_scale"},
}

// ActionDisplay reads or sets display settings.
func ActionDisplay(serial, model, key, value string) error {
	if key != "" && value != "" {
		key = strings.ToLower(key)
		if key == "dpi" {
			_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell", "wm", "density", value})
			if code != 0 {
				return nterrors.AdbError(fmt.Sprintf("wm density %s failed: %s", value, strings.TrimSpace(stderr)))
			}
			fmt.Printf("  DPI set to %s on %s.\n", value, model)
			return nil
		}
		ns, ok := displaySettings[key]
		if !ok {
			keys := []string{"dpi"}
			for k := range displaySettings {
				keys = append(keys, k)
			}
			return nterrors.AdbError(fmt.Sprintf("Unknown display key '%s'. Valid keys: %s", key, strings.Join(keys, ", ")))
		}
		if err := putSetting(serial, ns[0], ns[1], value); err != nil {
			return err
		}
		fmt.Printf("  %s set to %s on %s.\n", key, value, model)
		return nil
	}

	sizeRaw := adb.ShellStr(serial, "wm size")
	resolution := parseWMSize(sizeRaw)
	densityRaw := adb.ShellStr(serial, "wm density")
	dpi := parseWMDensity(densityRaw)
	displayDump := adb.ShellStr(serial, "dumpsys display | grep -E 'mRefreshRate|refreshRate'")
	refreshRate := parseRefreshRate(displayDump)

	brightness := getSetting(serial, "system", "screen_brightness")
	brightnessMode := getSetting(serial, "system", "screen_brightness_mode")
	timeoutRaw := getSetting(serial, "system", "screen_off_timeout")
	fontScale := getSetting(serial, "system", "font_scale")
	rotationVal := getSetting(serial, "system", "user_rotation")
	rotationAuto := getSetting(serial, "system", "accelerometer_rotation")

	brightnessDisplay := "n/a"
	if brightness != "" {
		brightnessDisplay = fmt.Sprintf("%s / 255 (auto: %s)", brightness, fmtOnOff(brightnessMode))
	}
	timeoutDisplay := "n/a"
	if timeoutRaw != "" {
		timeoutDisplay = fmtTimeout(timeoutRaw)
	}
	fontDisplay := fontScale
	if fontDisplay == "" {
		fontDisplay = "n/a"
	}
	rotationDisplay := fmtRotation(rotationVal)
	if rotationAuto != "" {
		rotationDisplay += fmt.Sprintf(" (auto: %s)", fmtOnOff(rotationAuto))
	}

	fmt.Printf("\n  Display \u2014 %s\n\n", model)
	fmt.Printf("  %-16s: %s\n", "Resolution", resolution)
	fmt.Printf("  %-16s: %s\n", "DPI", dpi)
	fmt.Printf("  %-16s: %s\n", "Refresh Rate", refreshRate)
	fmt.Println()
	fmt.Printf("  %-16s: %s\n", "Brightness", brightnessDisplay)
	fmt.Printf("  %-16s: %s\n", "Font Scale", fontDisplay)
	fmt.Printf("  %-16s: %s\n", "Rotation", rotationDisplay)
	fmt.Printf("  %-16s: %s\n", "Screen Timeout", timeoutDisplay)
	fmt.Println()
	return nil
}

var profileAliases = map[string]string{
	"natural": "0",
	"vivid":   "1",
	"custom":  "256",
}

// ActionColorProfile reads or sets the display color profile.
func ActionColorProfile(serial, model, profile string) error {
	if profile != "" {
		canonical, ok := profileAliases[strings.ToLower(profile)]
		if !ok {
			canonical = profile
		}
		if err := putSetting(serial, "system", "display_color_mode", canonical); err != nil {
			return err
		}
		label := colorProfileLabel(canonical)
		fmt.Printf("  Color profile set to %s on %s.\n", label, model)
		return nil
	}

	colorMode := getSetting(serial, "system", "display_color_mode")
	nightActive := getSetting(serial, "secure", "night_display_activated")
	nightTemp := getSetting(serial, "secure", "night_display_color_temperature")

	modeDisplay := "n/a"
	if colorMode != "" {
		modeDisplay = fmt.Sprintf("%s (%s)", colorProfileLabel(colorMode), colorMode)
	}
	nightDisplay := fmtOnOff(nightActive)

	fmt.Printf("\n  Color Profile \u2014 %s\n\n", model)
	fmt.Printf("  %-16s: %s\n", "Color Mode", modeDisplay)
	fmt.Printf("  %-16s: %s\n", "Night Light", nightDisplay)
	if nightActive == "1" && nightTemp != "" {
		fmt.Printf("  %-16s: %s K\n", "Color Temp", nightTemp)
	}
	fmt.Println()
	return nil
}

File: python/nothing/display.py
"""Display settings and color profile management for Nothing phones."""

import re

from .device import run
from .exceptions import AdbError
from .models import DeviceInfo


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _setting(serial: str, namespace: str, key: str) -> str:
    """Read an Android setting via 'adb shell settings get'."""
    r = run(["adb", "-s", serial, "shell", "settings", "get", namespace, key])
    val = r.stdout.strip() if r.returncode == 0 else ""
    return "" if val in ("null", "null\n", "null\r") else val


def _put_setting(serial: str, namespace: str, key: str, value: str) -> None:
    """Write an Android setting; raise AdbError on failure."""
    r = run(["adb", "-s", serial, "shell", "settings", "put", namespace, key, value])
    if r.returncode != 0:
        raise AdbError(f"settings put {namespace} {key} failed: {r.stderr.strip()}")


def _shell(serial: str, cmd: str) -> str:
    """Run adb shell command; return stdout stripped, empty string on failure."""
    r = run(["adb", "-s", serial, "shell", cmd])
    return r.stdout.strip() if r.returncode == 0 else ""


# ---------------------------------------------------------------------------
# Internal formatters
# ---------------------------------------------------------------------------

def _fmt_timeout(ms_str: str) -> str:
    """Convert a screen timeout value in milliseconds to a human-readable string."""
    try:
        ms = int(ms_str)
    except (ValueError, TypeError):
        return ms_str or "n/a"

    seconds = ms // 1000
    if seconds < 60:
        return f"{seconds}s"
    minutes = seconds // 60
    if minutes == 1:
        return "1 min"
    return f"{minutes} min"


def _fmt_rotation(val: str) -> str:
    """Convert user_rotation integer to a human-readable label."""
    mapping = {"0": "portrait", "1": "landscape", "2": "reverse-portrait", "3": "reverse-landscape"}
    return mapping.get(val, val or "n/a")


def _fmt_on_off(val: str) -> str:
    """Return 'on' or 'off' from a '1'/'0' setting value."""
    if val == "1":
        return "on"
    if val == "0":
        return "off"
    return val or "n/a"


def _parse_wm_size(raw: str) -> str:
    """Extract resolution string from 'wm size' output (e.g. 'Physical size: 1080x2400')."""
    m = re.search(r'Physical size:\s*(\d+x\d+)', raw)
    if m:
        return m.group(1) + " (physical)"
    m = re.search(r'(\d+x\d+)', raw)
    if m:
        return m.group(1)
    return raw or "n/a"


def _parse_wm_density(raw: str) -> str:
    """Extract DPI value from 'wm density' output."""
    # Output is either "Physical density: 420" or "Override density: 420"
    m = re.search(r'(?:Override|Physical) density:\s*(\d+)', raw)
    if m:
        return m.group(1)
    m = re.search(r'(\d+)', raw)
    if m:
        return m.group(1)
    return raw or "n/a"


def _parse_refresh_rate(raw: str) -> str:
    """Extract the first refresh rate value from dumpsys display output."""
    for line in raw.splitlines():
        # Match patterns like: mRefreshRate=120.0, refreshRate=120.0, "refreshRate": 120.0
        m = re.search(r'(?:mRefreshRate|refreshRate)[=:\s"]+([0-9]+(?:\.[0-9]+)?)', line)
        if m:
            return m.group(1) + " Hz"
    return "n/a"


def _color_profile_label(val: str) -> str:
    """Return a human-readable label for a Nothing OS display_color_mode value."""
    mapping = {"0": "Natural (sRGB)", "1": "Vivid (P3)", "256": "Custom"}
    return mapping.get(val, f"Unknown ({val})" if val else "n/a")


# ---------------------------------------------------------------------------
# Key → (namespace, settings_key) map for action_display writes
# ---------------------------------------------------------------------------

_DISPLAY_SETTINGS: dict[str, tuple[str, str]] = {
    "brightness":      ("system", "screen_brightness"),
    "brightness_auto": ("system", "screen_brightness_mode"),
    "timeout":         ("system", "screen_off_timeout"),
    "rotation":        ("system", "user_rotation"),
    "rotation_auto":   ("system", "accelerometer_rotation"),
    "font_scale":      ("system", "font_scale"),
}


# ---------------------------------------------------------------------------
# 1. Display settings
# ---------------------------------------------------------------------------

def action_display(device: DeviceInfo, key: str | None, value: str | None) -> None:
    """Read or set display settings on the connected Nothing phone.

    With no key/value: prints a summary of all display settings.
    With key + value:  writes the specified setting.
    """
    s = device.serial

    # ── Write mode ────────────────────────────────────────────────────────
    if key is not None and value is not None:
        key = key.lower()

        if key == "dpi":
            r = run(["adb", "-s", s, "shell", "wm", "density", value])
            if r.returncode != 0:
                raise AdbError(f"wm density {value} failed: {r.stderr.strip()}")
            print(f"  DPI set to {value} on {device.model}.")
            return

        if key not in _DISPLAY_SETTINGS:
            raise AdbError(
                f"Unknown display key '{key}'. "
                f"Valid keys: {', '.join(sorted(_DISPLAY_SETTINGS) + ['dpi'])}"
            )

        namespace, setting_key = _DISPLAY_SETTINGS[key]
        _put_setting(s, namespace, setting_key, value)
        print(f"  {key} set to {value} on {device.model}.")
        return

    # ── Read mode ─────────────────────────────────────────────────────────
    # Physical size
    size_raw     = _shell(s, "wm size")
    resolution   = _parse_wm_size(size_raw)

    # DPI — prefer override density if present, fall back to physical
    density_raw  = _shell(s, "wm density")
    dpi          = _parse_wm_density(density_raw)

    # Refresh rate — first match from dumpsys display
    display_dump = _shell(s, "dumpsys display | grep -E 'mRefreshRate|refreshRate'")
    refresh_rate = _parse_refresh_rate(display_dump)

    # Settings-based values
    brightness      = _setting(s, "system", "screen_brightness")
    brightness_mode = _setting(s, "system", "screen_brightness_mode")
    timeout_raw     = _setting(s, "system", "screen_off_timeout")
    font_scale      = _setting(s, "system", "font_scale")
    rotation_val    = _setting(s, "system", "user_rotation")
    rotation_auto   = _setting(s, "system", "accelerometer_rotation")

    # Formatted values
    brightness_display = brightness or "n/a"
    if brightness:
        brightness_display = f"{brightness} / 255 (auto: {_fmt_on_off(brightness_mode)})"

    timeout_display  = _fmt_timeout(timeout_raw) if timeout_raw else "n/a"
    font_display     = font_scale or "n/a"
    rotation_display = _fmt_rotation(rotation_val)
    if rotation_auto:
        rotation_display += f" (auto: {_fmt_on_off(rotation_auto)})"

    # ── Output ────────────────────────────────────────────────────────────
    print(f"\n  Display — {device.model}\n")
    print(f"  {'Resolution':<16}: {resolution}")
    print(f"  {'DPI':<16}: {dpi}")
    print(f"  {'Refresh Rate':<16}: {refresh_rate}")
    print()
    print(f"  {'Brightness':<16}: {brightness_display}")
    print(f"  {'Font Scale':<16}: {font_display}")
    print(f"  {'Rotation':<16}: {rotation_display}")
    print(f"  {'Screen Timeout':<16}: {timeout_display}")
    print()


# ---------------------------------------------------------------------------
# 2. Color profile
# ---------------------------------------------------------------------------

# Accepted profile name aliases → canonical integer string
_PROFILE_ALIASES: dict[str, str] = {
    "natural": "0",
    "vivid":   "1",
    "custom":  "256",
}


def action_color_profile(device: DeviceInfo, profile: str | None) -> None:
    """Read or set the display color profile on the connected Nothing phone.

    With no profile: prints current color mode and night light settings.
    With profile:    sets display_color_mode to the requested value.
    """
    s = device.serial

    # ── Write mode ────────────────────────────────────────────────────────
    if profile is not None:
        canonical = _PROFILE_ALIASES.get(profile.lower(), profile)
        _put_setting(s, "system", "display_color_mode", canonical)
        label = _color_profile_label(canonical)
        print(f"  Color profile set to {label} on {device.model}.")
        return

    # ── Read mode ─────────────────────────────────────────────────────────
    color_mode       = _setting(s, "system", "display_color_mode")
    night_active     = _setting(s, "secure", "night_display_activated")
    night_temp       = _setting(s, "secure", "night_display_color_temperature")

    mode_display     = _color_profile_label(color_mode) if color_mode else "n/a"
    if color_mode:
        mode_display = f"{_color_profile_label(color_mode)} ({color_mode})"

    night_display    = _fmt_on_off(night_active)

    # ── Output ────────────────────────────────────────────────────────────
    print(f"\n  Color Profile — {device.model}\n")
    print(f"  {'Color Mode':<16}: {mode_display}")
    print(f"  {'Night Light':<16}: {night_display}")
    if night_active == "1" and night_temp:
        print(f"  {'Color Temp':<16}: {night_temp} K")
    print()
